internal/subscriber/subscriberimpl_test.go

```diff
@@ -0,0 +1,301 @@
+			Data:      data,
+			Topic:     Topic,
+			Partition: Partition,
+			Offset:    input.offset,
+		}
+		msgProto.MessageID = input.messageID
+		messages = append(messages, msgProto)
+	}
+	cs.EXPECT().ReceiveMessages(ctx, gomock.Any()).Return(
+		&messagebroker.GetMessagesFromTopicResponse{
+			Messages: messages,
+		}, nil,
+	)
+
+	pullRequest := &PullRequest{ctx: ctx, MaxNumOfMessages: int32(len(testInputs))}
+	responseChan := make(chan *metrov1.PullResponse, 1)
+	errChan := make(chan error, 10)
+	subImpl.Pull(ctx, pullRequest, responseChan, errChan)
+
+	ticker := time.NewTimer(30 * time.Second)
+	select {
+	case resp := <-responseChan:
+		tp := NewTopicPartition(Topic, Partition)
+		assert.Equal(t, len(testInputs), len(resp.ReceivedMessages))
+		for index, msg := range resp.ReceivedMessages {
+			ackMsg, _ := ParseAckID(msg.AckId)
+			subImpl.Acknowledge(ctx, ackMsg, errChan)
+			assert.Equal(t, testInputs[index].canConsumeMore, subImpl.CanConsumeMore())
+			assert.Equal(t, testInputs[index].maxCommittedOffset, subImpl.consumedMessageStats[tp].maxCommittedOffset)
+		}
+	case <-ticker.C:
+		assert.FailNow(t, "Test case timed out")
+	}
+}
+
+func TestBasicImplementation_ModAckDeadline(t *testing.T) {
+	ctx, cs, subImpl := setup(t)
+	testInputs := []struct {
+		message              string
+		messageID            string
+		ackDeadline          int32
+		consumedMessageCount int
+	}{
+		{
+			message:              "Hello",
+			messageID:            "1",
+			ackDeadline:          int32(0),
+			consumedMessageCount: 1,
+		},
+		{
+			message:              "there",
+			messageID:            "2",
+			ackDeadline:          int32(time.Now().Add(1 * time.Nanosecond).Unix()),
+			consumedMessageCount: 1,
+		},
+	}
+
+	messages := make([]messagebroker.ReceivedMessage, 0, 3)
+	for _, input := range testInputs {
+		data, _ := proto.Marshal(&metrov1.PubsubMessage{Data: []byte(input.message)})
+		msgProto := messagebroker.ReceivedMessage{
+			Data:      data,
+			Topic:     Topic,
+			Partition: Partition,
+		}
+		msgProto.MessageID = input.messageID
+		messages = append(messages, msgProto)
+	}
+
+	cs.EXPECT().ReceiveMessages(ctx, gomock.Any()).Return(
+		&messagebroker.GetMessagesFromTopicResponse{
+			Messages: messages,
+		}, nil,
+	)
+
+	pullRequest := &PullRequest{ctx: ctx, MaxNumOfMessages: 1}
+	responseChan := make(chan *metrov1.PullResponse, 1)
+	errChan := make(chan error, 10)
+	subImpl.Pull(ctx, pullRequest, responseChan, errChan)
+
+	ticker := time.NewTimer(30 * time.Second)
+	select {
+	case resp := <-responseChan:
+		assert.NotEmpty(t, resp.ReceivedMessages)
+
+		tp := NewTopicPartition(Topic, Partition)
+		assert.Equal(t, len(testInputs), len(resp.ReceivedMessages))
+		for index, msg := range resp.ReceivedMessages {
+			ackMsg, _ := ParseAckID(msg.AckId)
+			modAckReq := &ModAckMessage{
+				ctx:         ctx,
+				AckMessage:  ackMsg,
+				ackDeadline: testInputs[index].ackDeadline,
+			}
+			subImpl.ModAckDeadline(ctx, modAckReq, errChan)
```

Comment:
Can we send a modack request over ModAckChan and process that instead?

Reply:
If I process it over the modAckChan, then the subscriber Run() needs to be called, which is in for loop. Not sure how to end that,

Reply:
Subscriber also has a `cancelFunc` that is part of the struct and can be invoked by calling `Subscriber.Stop()`